util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+        file = NULL;
+
+        if (!status.ok())
+        {
+            env_->DeleteFile(manifest_string);
+            Log(GetLogger(), "HotBackup failed writing/closing manifest (%s)",
+                status.ToString().c_str());
+            good=false;
+        }   // if
+    }   // if
+    else
+    {
+        Log(GetLogger(), "HotBackup failed creating manifest (%s)",
+            status.ToString().c_str());
+        good=false;
+    }   // else
+
+    // create CURRENT file that points to new manifest
+    if (good)
+    {
+        status = SetCurrentFile(env_, local_options.tiered_fast_prefix, manifest_num);
+
+        if (!status.ok())
+        {
+            env_->DeleteFile(manifest_string);
+            Log(GetLogger(), "HotBackup failed to set manifest in CURRENT (%s)",
+                status.ToString().c_str());
+            good=false;
+        }   // if
+    }   // if
+
+    // if the manifest is good, create file links
+    if (good)
+        good=CreateBackupLinks(version, local_options);
+
+    // success or failure, must release version (within mutex)
+    {
+        MutexLock l(&mutex_);
+        version->Unref();               // must have mutex for Ref/Unref
+    }   // mutex
+
+    return(good);
+
+}   // DBImpl::WriteBackupManifest
+
+
+/**
+ * Create a hard link between each live file and its backup name.
+ *  This forces necessary .sst files to survive in the backup directory
+ *  even if no longer needed in active directory.  Also duplicates only
+ *  the name, not contents ... saving tons of space.
+ */
+bool
+DBImpl::CreateBackupLinks(
+    Version * Ver,
+    Options & BackupOptions)
+{
+    bool good(true);
+    int level, ret_val;
+    std::string src_path, dst_path;
+
+    for (level=0; level<config::kNumLevels && good; ++level)
+    {
+        const Version::FileMetaDataVector_t level_files(Ver->GetFileList(level));
+        Version::FileMetaDataVector_t::const_iterator it;
+
+        for (it=level_files.begin(); it!=level_files.end() && good; ++it)
+        {
+            src_path=TableFileName(options_, (*it)->number, level);
+            dst_path=TableFileName(BackupOptions, (*it)->number, level);
+
+            // this would be cleaner if link operation added to Env class heirarchy
+            ret_val=link(src_path.c_str(), dst_path.c_str());
+            if (0!=ret_val)
+            {
+                good=false;
+                Log(GetLogger(), "HotBackup failed linking (return val %d) %s",
+                    ret_val, src_path.c_str());
+            }   // if
+        }   // for
+    }   // for
+
+    return(good);
+
+}   // DBImpl::CreateBackupLinks
+
+
+/**
+ * Copy portion of LOG file that most likely applies to this backup.
+ */
+bool
+DBImpl::CopyLOGSegment(long EndPos)
+{
+    assert(0<EndPos);
+
+    bool good(true);
+    long remaining;
+    Options local_options;
+    SequentialFile * src(NULL);
+    WritableFile * dst(NULL);
+    Status s;
+    std::string src_name, dst_name, buffer;
```

Comment:
Holy crap! I have never seen an implementation of std::vector<> that didn't allocate a contiguous memory block for the underlying storage, but apparently the C++98 spec doesn't preclude doing so.

I just found a discussion about this on stackoverflow (http://stackoverflow.com/questions/849168/are-stdvector-elements-guaranteed-to-be-contiguous). Apparently this omission was corrected in C++98-TC1, and the new C++11 standard makes this requirement explicit.

Without looking at the implementations of some of our "problem child" compilers (e.g., FreeBSD and Solaris), I can't say whether or not they do the right thing with vector allocation, so sticking with your std::string approach seems wise for now.